pkg: rngd service support on AArch64

arm64b opened the pull request:
Enable the rngd service in case of a real hardware RNG
device is available on AArch64 platform

Signed-off-by: Dennis Chen <dennis.chen@arm.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/linuxkit/linuxkit/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Enable the `rngd` service on AArch64 in case of a hardware RNG device available
**- How I did it**
Check the `/dev/hwrng` and use it as the entropy source to feed the `/dev/random` if `/dev/hwrng` is the real character device of a RNG hardware.
**- How to verify it**
1. `linuxkit pkg build -org m1130 rngd/`
2.  `docker run --privileged --rm --device /dev/hwrng m1130/rngd:842e5e8ece7934f0cab9fd0027b595ff3471e5b9-dirty`
3. On an AArch64 platform which has a real RNG device such as Cavium ThunderX, we can see the `/sbin/rngd` is running inside the docker.
**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**
![summit](https://user-images.githubusercontent.com/29669302/33416678-d033975e-d5d8-11e7-9c82-48a80afded73.jpg)

arm64b commented:
@justincormack @ijc @rn, currently we always pass a `virtio-rng` device backed by host's `/dev/urandom` to the guest, that makes the guest's `/dev/hwrng` always valid, in this case the entropy from the `/dev/hwrng` maybe not good enough for some security-sensitive applications. My proposal here is, if there's a real RNG hardware device in the host, we can para-virtualise that to the guest, else we do nothing, thus the `/dev/hwrng` in the guest will give the effective information of the quality of the entropy source. Any comments about that?

deitch commented:
@arm64b I am a bit confused by this. If I understood you correctly, the problem is as follows.

We always pass `/dev/urandom` of host to guest as `/dev/hwrng`. This makes the guest always _think_ that it has a real `hwrng`, when it may, in fact, be backed by a lower-entropy device.

Essentially, the expectation on Linux is:

* use `hwrng` to discover if you really have a hardware rng
* use `random` if you want reliable entropy and are willing to wait for it
* use `urandom` if you want some entropy, filled in with lower quality if there isn's sufficient available

By promoting `hwrng` in the guest to `urandom` in the host, we are "tricking" the guest into thinking there is a real entropy when there isn't.

Is that it?

arm64b commented:
@deitch Correct. With `virtio-rng`, we fake a RNG device in the guest but it actually backed by `/dev/urandom`, the latter doesn't block if it does not generate enough entropy, eg, output the same number repeatedly. The essential is we need to make sure guest realize this.

deitch commented:
In that case, I don't get how this PR helps. Doesn't this change `pkg/rngd` which runs _inside_ the guest and thus has no knowledge? Would we not need to change how `lkt run` calls hyperkit or qemu command-line?

arm64b commented:
That's different, this PR just uses the entropy from `/dev/hwrng` to feed `/dev/random` in guest, so some security applications can take use of `/dev/random` to generate like keys, the PR doesn't care about the entropy quality in `/dev/hwrng`. But if the guest can realize its `/dev/hwrng` is backed by a real RNG device or not, we can make better decision how to handle it.

deitch commented:
Ah, so you are not fixing that issue, but rather intelligently checking if `hwrng` exists in the guest? But again, won't it exist in the guest and be (possibly) fake anyways, because it is backed by `urandom` in the underlying host?

arm64b commented:
> so you are not fixing that issue

Right, that's another topic, it's my proposal.

> won't it exist in the guest and be (possibly) fake anyways, because it is backed by urandom in the underlying host?

For our current solution, the `/dev/hwrng` always valid, as you said, it's backed by `/dev/urandom` in the host. So this PR doesn't care about the backend of the `/dev/hwrng`, it just establishes a bridge between the `/dev/random` and `/dev/hwrng` in guest: read entropy from `/dev/hwrng` and feed it to `/dev/random`.

deitch commented:
> So this PR doesn't care about the backend of the /dev/hwrng, it just establishes a bridge between the /dev/random and /dev/hwrng in guest: read entropy from /dev/hwrng and feed it to /dev/random

And your plan then is to have a future PR that will replace the blind matching of `urandom` (host) -> `hwrng` (guest) with something more intelligent like `hwrng` (host if it exists) -> `hwrng` (guest)?

This would bring `rngd` in line with its usual job, which is `hwrng`/`hwrandom` -> `random`.

Looks good to me, but like to get input from another someone with lower-level understanding. I believe @justincormack did the original `rngd` implementation?

ijc commented on pkg/rngd/cmd/rngd/rng_linux_arm64.go:
```diff
@@ -7,15 +7,47 @@ package main
+		return 0, err
+	}
+
+	r = binary.LittleEndian.Uint64(buf)
```
OOI why is the endianess of the random data relevant?

arm64b replied:
Right. Will consolidate that with Justin's comments followed.

rn commented on pkg/rngd/cmd/rngd/rng_linux_arm64.go:
```diff
@@ -7,15 +7,47 @@ package main
+	f, err := os.Open("/dev/hwrng")
+	defer f.Close()
+	if err != nil {
+		log.Fatalf("Cannot open /dev/hwrng: %v", err)
```
`log.Fatalf` will exit so the return below will not happen and `rngd` will not run. What is the intention here? Exit or return?

arm64b replied:
Will return false indicate `rngd` will not be launched with `log.Fatalf` removed.

rn commented on pkg/rngd/cmd/rngd/rng_linux_arm64.go:
```diff
@@ -7,15 +7,47 @@ package main
+	buf := make([]byte, 16)
+	n, err := f.Read(buf)
+	if err != nil || n != 16 {
+		log.Fatalf("Read /dev/hwrng failed: %v", err)
```
ditto

justincormack commented:
We should not make this code arm64 specific - `/dev/hwrng` is generic and will cover a tpm rng on amd64 platform too, which I have been meaning to add support for, so we should move it to generic code that is used when this device exists.

rn commented on pkg/rngd/cmd/rngd/rng_linux_arm64.go:
```diff
@@ -7,15 +7,47 @@ package main
+	// Try to read some data from the entropy source. If it doesn't
+	// return an error, assume it's ok to use
+	buf := make([]byte, 16)
+	n, err := f.Read(buf)
```
Is this a read on `/dev/hwrng` blocking if there is no entropy, like `/dev/random`? If so we want a timeout on this,

arm64b replied:
I am not pretty sure since I can't test all the hardware RNG devices, but if that case happens, probably it's hardware bug.

justincormack commented on pkg/rngd/cmd/rngd/rng_linux_arm64.go:
```diff
@@ -7,15 +7,47 @@ package main
+	defer f.Close()
+	if err != nil {
+		log.Fatalf("Cannot open /dev/hwrng: %v", err)
+		return false
```
log.Fatalf will exit, so no point returning.

arm64b replied:
Yes

justincormack commented on pkg/rngd/cmd/rngd/rng_linux_arm64.go:
```diff
@@ -7,15 +7,47 @@ package main
+		return 0, err
+	}
+
+	r = binary.LittleEndian.Uint64(buf)
```
endianness of random data doesnt matter, it is still random bits.

arm64b replied:
Will consolidate with ijc's comment above

justincormack commented:
@arm64b in the virtual case, the passthrough that `linuxkit` does is not necessarily meant to be what you use in production, it is meant as a good developer default. Generally you should use qemu/kvm directly and decide what to pass depending on what the host rng situation is and what the requirements are.

deitch commented:
> Generally you should use qemu/kvm directly and decide what to pass depending on what the host rng situation is and what the requirements are

Probably still is a good thing to "do the right thing" when we run qemu and pass through matching devices, though. If @arm64b wants to put in the work to make that happen, probably a good thing, no?

In prod though, yes, definitely should set qemu command as you want (or maybe don't use qemu?)

justincormack commented:
@deitch I don't think in general you want to pass through hw rng, you probably want to mix them into host entropy and pass that. Eitehr way we should not be providing policy for production in the `linuxkit` tool, people should make those decisions themselves and use the tools directly.

deitch commented:
> you probably want to mix them into host entropy and pass that

Didn't understand this. What is mapped to what? Does `hwrng` exist on the guest and if so, mapped to where? If it is to `random` (or `urandom`), then it is not a true `hwrng`.

Or are you saying `lkt run` isn't meant for production use, ever, so it doesn't matter?

deitch commented:
> I am also slightly confused about /dev/hwrng in the case of virtio rng as it seems to exist and work, but in fact the kernel talks to it directly and AFAIK rngd does not need to read it from userspace, but I am not 100% sure about this, need to look into it.

Oh, so it isn't just me who is confused? I feel better now.

I always thought `rngd`'s job was to read from a hardware random device (like `/dev/hwrng`) and feed `/dev/random`. Perhaps not?

arm64b commented:
> I am also slightly confused about /dev/hwrng in the case of virtio rng as it seems to exist and work, but in fact the kernel talks to it directly and AFAIK rngd does not need to read it from userspace, but I am not 100% sure about this, need to look into it.

IMO, /dev/hwrng generate the entropy saved in kernel space, while the entropy pool that  `/dev/random` used is also in kernel space, technically we can transmit the entropy generated by `/dev/hwrng` into the entropy pool of `/dev/random`, but actually there isn't a channel to do that in current kernel implementation, that's why we need the `rngd` to do that.

arm64b commented:
> I always thought rngd's job was to read from a hardware random device (like /dev/hwrng) and feed /dev/random. Perhaps not?

No. Kernel doesn't have such a data tx channel between `/dev/hwrng` and `/dev/random`, so we need a service like `rngd` to read entropy from `/dev/hwrng` and feed it into the entropy pool of `/dev/random`, the latter will be used by security-sensitive applications.

deitch commented:
> we need a service like rngd to read entropy from /dev/hwrng and feed it into the entropy pool of /dev/random

Uh, isn't that precisely what I asked?

> rngd's job was to read from a hardware random device (like /dev/hwrng) and feed /dev/random.

arm64b commented:
> Uh, isn't that precisely what I asked?

Ah, so I think your thought is correct 😄

arm64b commented:
Hello @justincormack @ijc @rn and @deitch , I just forced an update of this PR. In this PR, I addressed some comments from you guys and the important is to move the hardware RNG code as arch agnostic. This PR also has a framework to support both hardware RNG device and some RNG instruction such as it does on x86_64.  For endlessness, since the random number is not related with that, I still use it since this will make the code easy, but you guys have some better one to do that, I am happy to embrace that 😸

arm64b commented:
Hi Maintainers, I didn't get any feedback for the latest update, so don't know if it's ready or something still need to be changed. Would you please give some comments so I know which way to go. Thanks!
/cc @justincormack

deitch commented:
> I didn't get any feedback for the latest update

Mix of everyone either busy or at KubeCon (not I, sigh...). Or both.

Going to reread the code changes.

deitch commented:
Took some time to figure out what this is doing. I do have some comments, but they are mostly stylistic, so I will wait for the responses from the others involved.

A few inline, though.

deitch commented on pkg/rngd/cmd/rngd/main.go:
```diff
@@ -14,7 +15,76 @@ import (
+type rng struct {
+	rngName  string
+	rngDev   string
+	disabled bool
```
Should this not be `enabled`, so the default is false, and then you can enable it when an `init` succeeds?

arm64b replied:
Yes. enabled==true means a valid entropy source which can be used.

deitch replied:
Got it. Just a nitpick, but it is the reverse now, because the default for `disabled` is `false`

arm64b replied:
Ah, the code will set `disabled` as `true` for an invalid/malfunction entropy source, so when try to write the entropy to the pool, we have below code:
```
if e.disabled {
     continue
}
```
My comments is confusing 😄

deitch replied:
As I said, "just a nitpick". :-)

deitch commented on pkg/rngd/cmd/rngd/main.go:
```diff
@@ -14,7 +15,76 @@ import (
+}
+
 func main() {
+	var entSource int
```
Why is this an `int`? Why not a `bool` which defaults to `false`, and then you can set `entSource = true` when you have one at L98? You don't use the count elsewhere.

arm64b replied:
entSource stands for the total number of valid/enabled entropy sources in the system, if we use a bool variable, think about below sequence:
Source A: enabled      entSource=true
Source B: disabled     entSource=false
So we can't flip this variable back and forth.

arm64b replied:
Yes, make sense.

deitch commented on pkg/rngd/cmd/rngd/main.go:
```diff
@@ -14,7 +15,76 @@ import (
+		rngDev:  "/dev/hwrng",
+		rngInit: initEntropySource,
+		rngRead: readHwRNG,
+	},
```
Does this mean that it tries to read the generic (i.e. cross-platform) `/dev/hwrng` first, and thus will work on arm64 and amd64 if available, and only after try `DRNG`, which is Intel-specific?

arm64b replied:
Right, First we try to use hardware RNG device, then the arch specific method such as RNG instruction if that arch provides like amd64 does.

deitch commented on pkg/rngd/cmd/rngd/main.go:
```diff
@@ -81,16 +155,29 @@ type randInfo struct {
+	var r uint64
+	var err error
+	var ret uintptr
+	for _, e := range entropySources {
```
So, it takes from all enabled sources?

arm64b replied:
Yes, The more entropy sources available, the more better quality of the entropies in the pool, the more safer to use it.

justincormack commented:
Sorry, been travelling to Kubecon, will try to review soon!

justincormack commented on pkg/rngd/cmd/rngd/main.go:
```diff
@@ -14,7 +15,76 @@ import (
+		rngInit: initEntropySource,
+		rngRead: readHwRNG,
+	},
+	// TPM support will be here
```
I believe tmp uses `/dev/hwrng` too, so this comment is not needed

deitch replied:
I believe you are right.

arm64b replied:
OK, I will remote that.

justincormack commented on pkg/rngd/cmd/rngd/rng_linux_arm64.go:
```diff
@@ -10,12 +10,12 @@ import (
-
-func initRand() bool {
+// No standard RNG instruction on arm64
+func initDRNG(ctx *rng) bool {
```
maybe we should just change this file to be compiled when  `!linux,!amd64` as this will then be fine if we add other platforms or someone else wants to reuse the code.

arm64b replied:
I don't quite understand, if we use `!linux,!amd` in this file to prevent it compiling, then we will get `undefined initDRNG and readDRNG` error in 'linux+arm64' environment, so my alternative is: if we want to support other platforms, e.g. PowerPC, we can add a file like `rng_linux_ppc.go` to support RNG instruction case if that arch does have one.

arm64b commented:
Hi Guys,

How about the CI test failure:
...
[+ 77m 46s] Executing busybox-1.26.2-r7.trigger
[+ 77m 46s] OK: 319 MiB in 33 packages
[+ 77m 49s] Error processing tar file(exit status 1): write /usr/lib/go/pkg/tool/linux_amd64/pprof: no space left on device
[+ 77m 49s] exit status 1
[+ 77m 49s] Makefile:11: recipe for target 'build' failed
...
I am pretty sure the `rngd` pkg has been built successfully. `no space left on device` for what? 😢

arm64b commented:
/cc Hello @justincormack, I will be in long vacation for the upcoming Christmas & New year, so can we move forward for this PR and so we can have a very happy vacation together 🥇 ?

arm64b commented:
Seems a stale PR, since we don't need `rngd` to feed entropy to `virtio-rng` device, then let's close it.
